expression/builtin_convert_charset.go

```diff
@@ -256,26 +256,97 @@ func BuildFromBinaryFunction(ctx sessionctx.Context, expr Expression, tp *types.
+	funcPropNone: {
+		/* args != strings */
+		ast.Bin, ast.CharFunc, ast.DateFormat, ast.Oct, ast.Space,
+		/* only 1 string arg, no implicit conversion */
+		ast.CharLength, ast.CharacterLength, ast.FromBase64, ast.Lcase, ast.Left, ast.LoadFile,
+		ast.Lower, ast.LTrim, ast.Mid, ast.Ord, ast.Quote, ast.Repeat, ast.Reverse, ast.Right,
+		ast.RTrim, ast.Soundex, ast.Substr, ast.Substring, ast.Ucase, ast.Unhex, ast.Upper, ast.WeightString,
+		/* args are independent, no implicit conversion */
+		ast.Elt,
+	},
+	funcPropBinAware: {
+		/* result is binary-aware */
+		ast.ASCII, ast.BitLength, ast.Hex, ast.Length, ast.OctetLength, ast.ToBase64,
+		/* encrypt functions */
+		ast.AesDecrypt, ast.Decode, ast.Encode, ast.PasswordFunc, ast.MD5, ast.SHA, ast.SHA1,
+		ast.SHA2, ast.Compress, ast.AesEncrypt,
+	},
+	funcPropAuto: {
+		/* string functions */ ast.Concat, ast.ConcatWS, ast.ExportSet, ast.Field, ast.FindInSet,
+		ast.InsertFunc, ast.Instr, ast.Lpad, ast.Locate, ast.Lpad, ast.MakeSet, ast.Position,
+		ast.Replace, ast.Rpad, ast.SubstringIndex, ast.Trim,
+		/* operators */
+		ast.GE, ast.LE, ast.GT, ast.LT, ast.EQ, ast.NE, ast.NullEQ, ast.If, ast.Ifnull, ast.In,
+		ast.Case,
+		/* string comparing */
+		ast.Like, ast.Strcmp,
+		/* regex */
+		ast.Regexp,
+	},
+}
+
+var convertFuncsMap = map[string]funcProp{}
+
+func init() {
+	for k, fns := range convertActionMap {
+		for _, f := range fns {
+			convertFuncsMap[f] = k
+		}
+	}
+}
+
 // HandleBinaryLiteral wraps `expr` with to_binary or from_binary sig.
 func HandleBinaryLiteral(ctx sessionctx.Context, expr Expression, ec *ExprCollation, funcName string) Expression {
-	switch funcName {
-	case ast.Concat, ast.ConcatWS, ast.Lower, ast.Lcase, ast.Reverse, ast.Upper, ast.Ucase, ast.Quote, ast.Coalesce,
-		ast.Left, ast.Right, ast.Repeat, ast.Trim, ast.LTrim, ast.RTrim, ast.Substr, ast.SubstringIndex, ast.Replace,
-		ast.Substring, ast.Mid, ast.Translate, ast.InsertFunc, ast.Lpad, ast.Rpad, ast.Elt, ast.ExportSet, ast.MakeSet,
-		ast.FindInSet, ast.Regexp, ast.Field, ast.Locate, ast.Instr, ast.Position, ast.GE, ast.LE, ast.GT, ast.LT, ast.EQ,
-		ast.NE, ast.NullEQ, ast.Strcmp, ast.If, ast.Ifnull, ast.Like, ast.In, ast.DateFormat, ast.TimeFormat:
-		if ec.Charset == charset.CharsetBin && expr.GetType().Charset != charset.CharsetBin {
+	argChs, dstChs := expr.GetType().Charset, ec.Charset
+	switch convertFuncsMap[funcName] {
+	case funcPropNone:
+		return expr
+	case funcPropBinAware:
+		if isLegacyCharset(argChs) {
+			return expr
+		}
+		return BuildToBinaryFunction(ctx, expr)
+	case funcPropAuto:
+		if argChs != charset.CharsetBin && dstChs == charset.CharsetBin {
+			if isLegacyCharset(argChs) {
+				return expr
+			}
 			return BuildToBinaryFunction(ctx, expr)
-		} else if ec.Charset != charset.CharsetBin && expr.GetType().Charset == charset.CharsetBin {
+		} else if argChs == charset.CharsetBin && dstChs != charset.CharsetBin {
+			if ctx.GetSessionVars().SkipUTF8Check && isLegacyCharset(dstChs) {
```

Comment:
`SkipUTF8Check` is only used when insert data, why we should consider this variable?

Reply:
LegacyCharset should also wrap with `frombinary` for example 
```sql
mysql> select concat(0x91, 'a' collate utf8mb4_bin);
ERROR 3854 (HY000): Cannot convert string '\x91' from binary to utf8mb4
```